### PR TITLE
Update top-level haddocks for consensus

### LINKED ISSUE #4149: Provide an hyperlinked map of consensus modules
## Context 

* graphmod allows to display the dependencies among Haskell modules as a graph
* We want to provide a useful "map" of the consensus code base, displaying key modules/components and their relationships
* A map should help to navigate hence provide links to important parts of the system

## Goal

* goal of this task is to use existing graphmod script to generate an hyperlinked map of consensus modules grouped by components

### PR DESCRIPTION
**Pull request by abailly-iohk**
# Description

Fix #4149 #4057

This PR introduces a section about `Consensus Components` in the prolog of generated haddock documentation, and add a couple diagrams to try to provide a better overview of what's going on in the consensus codebase.

Screenshot 2022-11-29 at 12 30 37

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: scripts/prolog [hunk 1]
```diff
@@ -7,60 +7,77 @@ This site contains Haskell documentation of
...
-    * __[Win32-network](/Win32-network)__
-    * __[typed-protocols](typed-protocols/Network-TypedProtocol.html)__
-    * __[network-mux](network-mux/Network-Mux.html)__
+          * __[ChainSync](ouroboros-network/Ouroboros-Network-Protocol-ChainSync-Type.html)__
+          * __[BlockFetch](ouroboros-network/Ouroboros-Network-Protocol-BlockFetch-Type.html)__
+          * __[TxSubmission](ouroboros-network/Ouroboros-Network-Protocol-TxSubmission-Type.html)__
+          * __[TxSubmission2](ouroboros-network/Ouroboros-Network-Protocol-TxSubmission2-Type.html)__
+          * __[LocalTxSubmission](ouroboros-network/Ouroboros-Network-Protocol-LocalTxSubmission-Type.html)__
+          * __[LocalStateQuery](ouroboros-network/Ouroboros-Network-Protocol-LocalStateQuery-Type.html)__
+          * __[LocalTxMonitor](ouroboros-network/Ouroboros-Network-Protocol-LocalTxMonitor-Type.html)__
```
Comment (coot) [inline scripts/prolog:29]:
```suggestion
          * __[ChainSync](ouroboros-network-protocols/Ouroboros-Network-Protocol-ChainSync-Type.html)__
          * __[BlockFetch](ouroboros-network-protocols/Ouroboros-Network-Protocol-BlockFetch-Type.html)__
          * __[TxSubmission](ouroboros-network-protocols/Ouroboros-Network-Protocol-TxSubmission-Type.html)__
          * __[TxSubmission2](ouroboros-network-protocols/Ouroboros-Network-Protocol-TxSubmission2-Type.html)__
          * __[LocalTxSubmission](ouroboros-network-protocols/Ouroboros-Network-Protocol-LocalTxSubmission-Type.html)__
          * __[LocalStateQuery](ouroboros-network-protocols/Ouroboros-Network-Protocol-LocalStateQuery-Type.html)__
          * __[LocalTxMonitor](ouroboros-network-potocols/Ouroboros-Network-Protocol-LocalTxMonitor-Type.html)__
```

I forgot to update the links.

File: scripts/prolog [hunk 2]
```diff
@@ -7,60 +7,77 @@ This site contains Haskell documentation of
...
+     * /Block chain related data structures that the networking code needs
+       to be aware of:/
 
-    * __ouroboros-network-framework__ - our instantiations and implementation
-      of servers, clients, abstractions for working with named pipes and unix
-      sockets.  These are higher level components, which don't dependent on
-      ouroboros protocols, but build a solid foundation for the networking
-      library
+          * __[Block](ouroboros-network/Ouroboros-Network-Block.html)__
+          * __[AnchoredFragment](ouroboros-network/Ouroboros-Network-AnchoredFragment.html)__
```
Comment (coot) [inline scripts/prolog:36]:
```suggestion
          * __[Block](ouroboros-network-api/Ouroboros-Network-Block.html)__
          * __[AnchoredFragment](ouroboros-network-api/Ouroboros-Network-AnchoredFragment.html)__
```

File: scripts/prolog [hunk 3]
```diff
@@ -7,60 +7,77 @@ This site contains Haskell documentation of
...
-              * __[Data Diffusion](ouroboros-network/Ouroboros-Network-Diffusion.html)__ which combines all the pieces for supporting
-                a [node](https://github.com://input-output-hk/cardano-node)
-              * __[NodeToNode API](ouroboros-network/Ouroboros-Network-NodeToNode.html)__
-              * __[NodeToClient API](ouroboros-network/Ouroboros-Network-NodeToClient.html)__
-      
+* __ouroboros-consensus__ : Main package for Cardano consensus implementation, aka. Ouroboros.
+
+    * The [Ouroboros Consensus Modules Map](./modules-consensus.svg) represents the dependencies between various /Consensus/ modules and is a map to navigate the (large) codebase. Clicking on a module should bring you to the corresponding haddock page from which you can later navigate to the source code. __Note__: The image needs to be enlarged to be useful hence why it's an external link.
+
+* __ouroboros-consensus-test__ : Property and model-based tests for consensus, using [io-sim](https://)
```
Comment (coot) [inline scripts/prolog:62]:
```suggestion
* __ouroboros-consensus-test__ : Property and model-based tests for consensus, using [io-sim](https://github.com/input-output-hk/io-sim)
```

We will need to update it once it lands on `Hackage`.

File: scripts/prolog [hunk 4]
```diff
@@ -7,60 +7,77 @@ This site contains Haskell documentation of
...
+
+    * __[Network Simulator](ouroboros-consensus-test/Test-ThreadNet-Network.html)__: Simulates a network of nodes
```
Comment (coot) [inline scripts/prolog:64]:
```suggestion
    * __[ThreadNet Simulator](ouroboros-consensus-test/Test-ThreadNet-Network.html)__: Simulates a network of nodes
```

There's another thing which we refer to as a [network simulator](https://input-output-hk.github.io/ouroboros-network/ouroboros-network-framework/Simulation-Network-Snocket.html#v:withSnocket).

**Review by coot — CHANGES_REQUESTED** [review]

**Comment by abailly-iohk** [comment]
@coot I removed previous existing sections referring to io-sim and win32-network. Should have been another PR or a dedicated commit but 🤷

**Comment by abailly-iohk** [comment]
I have addressed the requested changes and I am reluctant to start working on something else before this PR gets merged. @coot @dnadales WDYT?

File: scripts/haddocs.sh [hunk 5]
```diff
@@ -92,3 +100,8 @@ for file in $(ls $OUTPUT_DIR/*/doc-index.json); do
     > /tmp/doc-index.json
   mv /tmp/doc-index.json "${OUTPUT_DIR}/doc-index.json"
 done
+
+# Copy modules map to output directory
+# TODO: dynamically generate
```
Comment (dnadales) [inline scripts/haddocs.sh:105]: Would it make sense to create an issue and address it rather soon?

**Review by dnadales — APPROVED** [review]
Looks great! Thank you.

File: scripts/haddocs.sh (hunk 5 above)
Comment (abailly-iohk) [reply scripts/haddocs.sh:105]: Definitely.

**Comment by abailly-iohk** [comment]
@coot @dnadales It seems this PR is blocked and has failing CI tasks which is annoying. I had a look at those failures and they seem completely unrelated to my changes, but having a red CI is definitely a blocker. I will try to rebase on master to see if that changes anything,

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**